sdk/python/lib/pulumi/automation/_cmd.py

```diff
@@ -35,46 +42,259 @@ def __str__(self) -> str:
+        script = tempfile.NamedTemporaryFile(delete=False)
+        try:
+            _downloadToFile("https://get.pulumi.com/install.sh", script.name)
+            os.chmod(script.name, 0o700)
+            script.close()  # The file was opened for writing, so we need to close it before executing it.
+            subprocess.check_output(
+                [
+                    script.name,
+                    "--no-edit-path",
+                    "--install-root",
+                    root,
+                    "--version",
+                    str(version),
+                ],
+                stderr=subprocess.STDOUT,
+            )
+        finally:
+            os.remove(script.name)
+
+    def run(
+        self,
+        args: List[str],
+        cwd: str,
+        additional_env: Mapping[str, str],
+        on_output: Optional[OnOutput] = None,
+    ) -> CommandResult:
+        """
+        Runs a Pulumi command, returning a CommandResult. If the command fails, a CommandError is raised.
+
+        :param args: The arguments to pass to the Pulumi CLI, for example `["stack", "ls"]`.
+        :param cwd: The working directory to run the command in.
+        :param additional_env: Additional environment variables to set when running the command.
+        :param on_output: A callback to invoke when the command outputs data.
+        """
+
+        # All commands should be run in non-interactive mode.
+        # This causes commands to fail rather than prompting for input (and thus hanging indefinitely).
+        if "--non-interactive" not in args:
+            args.append("--non-interactive")
+        env = {**os.environ, **additional_env}
+        if os.path.isabs(self.command):
+            env = _fixup_path(env, os.path.dirname(self.command))
+        cmd = ["pulumi"]
+        cmd.extend(args)
+
+        stdout_chunks: List[str] = []
+
+        with tempfile.TemporaryFile() as stderr_file:
+            with subprocess.Popen(
+                cmd, stdout=subprocess.PIPE, stderr=stderr_file, cwd=cwd, env=env
+            ) as process:
+                assert process.stdout is not None
+                while True:
+                    output = process.stdout.readline().decode(encoding="utf-8")
+                    if output == "" and process.poll() is not None:
+                        break
+                    if output:
+                        text = output.rstrip()
+                        if on_output:
+                            on_output(text)
+                        stdout_chunks.append(text)
+
+                code = process.returncode
+
+            stderr_file.seek(0)
+            stderr_contents = stderr_file.read().decode("utf-8")
+
+        result = CommandResult(
+            stderr=stderr_contents, stdout="\n".join(stdout_chunks), code=code
+        )
+        if code != 0:
+            raise create_command_error(result)
+
+        return result
+
+
+def _downloadToFile(url: str, path: str):
```

Comment:
Should this be `snake_case`? Same with the `install...` methods.

Reply:
Yes, thanks 😊